src/types/timezone.jl

```diff
@@ -1,4 +1,30 @@
-const TIME_ZONE_CACHE = Dict{String,Tuple{TimeZone,Class}}()
+# ---------- Time Zone Cache setup ----------------------------------------------------------------
+# Thread-local timezone caches to support multithreaded tasks simultaneously constructing TimeZones.
+# See https://github.com/JuliaTime/TimeZones.jl/issues/342 for more information.
+
+const THREADLOCAL_TIME_ZONE_CACHES = Dict{String,Tuple{TimeZone,Class}}[]
+
+@inline default_tz_cache() = default_tz_cache(Threads.threadid())
+@noinline function default_tz_cache(tid::Int)
+    0 < tid <= length(THREADLOCAL_TIME_ZONE_CACHES) || _tz_cache_length_assert()
+    if @inbounds isassigned(THREADLOCAL_TIME_ZONE_CACHES, tid)
+        @inbounds TZ_CACHE = THREADLOCAL_TIME_ZONE_CACHES[tid]
+    else
+        TZ_CACHE = eltype(THREADLOCAL_TIME_ZONE_CACHES)()
+        @inbounds THREADLOCAL_TIME_ZONE_CACHES[tid] = TZ_CACHE
+    end
+    return TZ_CACHE
+end
+@noinline _tz_cache_length_assert() =  @assert false "0 < tid <= length(THREADLOCAL_TIME_ZONE_CACHES)"
+
+module __InitThreadLocalCaches  # This module is only here to house the __init__() function.
+using ..TimeZones: THREADLOCAL_TIME_ZONE_CACHES
+function __init__()
+    resize!(empty!(THREADLOCAL_TIME_ZONE_CACHES), Threads.nthreads()) # ensures that we didn't save a bad object
```

Comment:
Why not just pre-allocated the empty dicts as part of the initialization?

Reply:
Yeah that's what I proposed originally in my comment, too, but @kristofferC said not to in his comment here:
https://github.com/JuliaTime/TimeZones.jl/issues/342#issuecomment-848501176

He says it's better to allocate the object on the thread where it will be used. Maybe it's for like memory locality or something? 🤔 

So anyway, i copied from the examples he posted. @kristofferC: Did I copy it right?

Reply:
The code you pasted did the allocation on one thread. Something like this should be fine:

```julia
@threads for i in 1:nthreads()
    THREADLOCAL_TIME_ZONE_CACHES[i] = Dict{String,Tuple{TimeZone,Class}}()
end

Reply:
Oh, interesting. Seems reasonable. In @kristofferC's example, his PR was explicitly titled "lazily allocate the threaded buffers and allocate them on the thread that will access it," so i figured there might be benefit to both aspects? What you suggest seems cleaner to me, too.

Reply:
After more thought the reason they are probably doing the lazy allocation is to avoid latency during package startup